examples/LaravelSampleApp/app/Http/Controllers/GoogleAdsApiController.php

```diff
@@ -0,0 +1,199 @@
+use Google\Ads\GoogleAds\Util\FieldMasks;
+use Google\Ads\GoogleAds\Util\V3\ResourceNames;
+use Google\Ads\GoogleAds\V3\Enums\CampaignStatusEnum\CampaignStatus;
+use Google\Ads\GoogleAds\V3\Resources\Campaign;
+use Google\Ads\GoogleAds\V3\Services\CampaignOperation;
+use Google\Ads\GoogleAds\V3\Services\GoogleAdsRow;
+use Google\Ads\GoogleAds\V3\Services\SearchGoogleAdsStreamResponse;
+use Illuminate\Http\Request;
+use Illuminate\Pagination\LengthAwarePaginator;
+use Illuminate\View\View;
+
+class GoogleAdsApiController extends Controller
+{
+    private static $REPORT_TYPE_TO_DEFAULT_SELECTED_FIELDS = [
+        'campaign' => ['campaign.id', 'campaign.name', 'campaign.status'],
+        'customer' => ['customer.id']
+    ];
+
+    /**
+     * Controls a POST or GET request submitted in the contect of the "Show Report" form.
+     *
+     * @param Request $request the HTTP request
+     * @param GoogleAdsClient $googleAdsClient the Google Ads API client
+     * @return View the view to redirect to after processing
+     */
+    public function showReportAction(
+        Request $request,
+        GoogleAdsClient $googleAdsClient
+    ): View {
+        if ($request->method() === 'POST') {
+            // Retrieves the form inputs.
+            $clientCustomerId = $request->input('clientCustomerId');
+            $reportType = $request->input('reportType');
+            $reportRange = $request->input('reportRange');
+            $entriesPerPage = $request->input('entriesPerPage');
+
+            // Retrieves the list of metric fields to select filtering out the static ones.
+            $selectedFields = array_values(
+                $request->except(
+                    [
+                        '_token',
+                        'clientCustomerId',
+                        'reportType',
+                        'reportRange',
+                        'entriesPerPage'
+                    ]
+                )
+            );
+
+            // Merges the list of metric fields to the resource ones that are selected by default.
+            $selectedFields = array_merge(
+                self::$REPORT_TYPE_TO_DEFAULT_SELECTED_FIELDS[$reportType],
+                $selectedFields
+            );
+
+            // Builds the GAQL query.
+            $query = sprintf(
+                "SELECT %s FROM %s WHERE metrics.impressions > 0 AND segments.date " .
+                "DURING %s LIMIT 100",
+                join(", ", $selectedFields),
+                $reportType,
+                $reportRange
+            );
+
+            // Searches the results using streaming.
+            /** @var GoogleAdsServerStreamDecorator $stream */
+            $stream = $googleAdsClient->getGoogleAdsServiceClient()->searchStream(
+                $clientCustomerId,
+                $query
+            );
+
+            // Fetches all the results.
+            $results = [];
+            foreach ($stream->readAll() as $response) {
```

Comment:
We cannot use the interceptor to hide a nested loop here?

Reply:
This is not released in 3.1.0. I will have to wait for 3.2.0 to make the change.